File: MtxVec.Delphi/Optim_Bounded.pas
unit Optim_Bounded;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Controls, Forms,
  Dialogs, Basic2, StdCtrls, ComCtrls, ExtCtrls, Grids, MtxExpr, Optimization,
  MtxIntDiff, MtxParseClass, MtxParseExpr, Math387;

type
  TfrmOptimBounded = class(TBasicForm2)
    Label1: TLabel;
    EditFunction: TEdit;
    Label2: TLabel;
    StringGridVars: TStringGrid;
    ButtonDetect: TButton;
    Button1: TButton;
    Memo1: TMemo;
    GroupBox1: TGroupBox;
    Label3: TLabel;
    Edit1: TEdit;
    Label4: TLabel;
    Edit2: TEdit;
    UpDown1: TUpDown;
    CheckBox1: TCheckBox;
    StringGridResults: TStringGrid;
    Label5: TLabel;
    procedure ButtonDetectClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure FormDestroy(Sender: TObject);
    procedure Button1Click(Sender: TObject);
    procedure EditFunctionChange(Sender: TObject);
  private
    { Private declarations }
    iVariables: TVariableCollection;
    pars,Lb,UB: Array of double;
    procedure InitSystem;
    procedure SolveProblem;
  public
    { Public declarations }
  end;


var
  frmOptimBounded: TfrmOptimBounded;

implementation

{$IFDEF CLR}
  {$R *.NFM}
{$ELSE}
  {$R *.DFM}
{$ENDIF}

uses MtxVec;

// A bit tricky, but ObjConsts hold all necessary objects, strings, etc..
// needed for calculation of f(x)
// ObjConsts[0] = TMtxExpression
// ObjConsts[1] = iVariables
// Parameters = parameter values
function OptObjFun(const Parameters: TVec; const Constants: TVec; Const ObjConst: Array of TObject): double;
var i: Integer;
  vars: TVariableCollection;
begin
  vars := ObjConst[1] as TVariableCollection;
  with ObjConst[0] as TMtxExpression do
  begin
    for i := 0 to Parameters.Length-1 do
    begin
      VarByName[vars[i].VarName].DoubleValue := Parameters.Values[i];
    end;
    Result := EvaluateDouble;
  end;
end;

procedure TfrmOptimBounded.Button1Click(Sender: TObject);
begin
  SolveProblem;
end;

procedure TfrmOptimBounded.ButtonDetectClick(Sender: TObject);
begin
  InitSystem;
  Button1.Enabled := true;
end;

procedure TfrmOptimBounded.EditFunctionChange(Sender: TObject);
begin
  Button1.Enabled := false;
end;




procedure TfrmOptimBounded.FormCreate(Sender: TObject);
begin
  inherited;
  EditFunction.Text := '100*Sqr(b-Sqr(a)) + Sqr(1-a)';
  iVariables := TVariableCollection.Create;
  With RichEdit1.Lines do
  begin
    Clear;
    Add('Optimization now supports bounded constraints. Invididual parameter '
      + 'can have no, lower, upper or both limits. This is done by:');
    Add('1. No limit: lower bound is -INF, upper bound is INF.');
    Add('2. Lower bound: lower bound is lb, upper bound is INF.');
    Add('3. Upper bound: lower bound is -INF, upper bound is ub.');
    Add('4. Lower and upper bound: lower bound is lb, upper bound is ub.');
    Add('');
    Add('1. Enter any function in the "Functon" edit box.');
    Add('2. Click the "Detect" button to get the list of variables in function.');
    Add('3. Define bounds (if any) for individual variables.');
    Add('4. Define starting values for individual variables.');
    Add('5. Press the "Calculate" button to minimize function.');
  end;


  Edit1.Text := SampleToStr(1e-8);
  InitSystem;
  Button1.Enabled := true;     

end;

procedure TfrmOptimBounded.FormDestroy(Sender: TObject);
begin
  iVariables.Free;
  SetLength(pars,0);
  SetLength(LB,0);
  SetLength(UB,0);
  pars := nil;
  LB := nil;
  UB := nil;
  inherited;
end;

procedure TfrmOptimBounded.InitSystem;
var list: TStrings;
  i: Integer;
  v: TNamedVariable;
  expr: TMtxExpression;
begin
  iVariables.Clear;
  list := TStringList.Create;
  expr := TMtxExpression.Create;
  try
    expr.AddExpr(EditFunction.Text);
    expr.GetVarList(list);

    StringGridVars.RowCount := List.Count + 1;
    StringGridResults.RowCount := List.Count + 1;
    StringGridVars.Cells[0,0] := 'Name';
    StringGridVars.Cells[1,0] := 'Lower bound';
    StringGridVars.Cells[2,0] := 'Upper bound';
    StringGridVars.Cells[3,0] := 'Initial value';
    StringGridResults.Cells[0,0] := 'Name';
    StringGridResults.Cells[1,0] := 'Value';

    for i := 0 to List.Count - 1 do
    begin
        v := iVariables.Add(List[i]);
        v.Value.DefineDouble;
    end;

    // Populate string grid
    for i := 0 to iVariables.Count-1 do
    begin
      StringGridVars.Cells[0,i+1] := iVariables.Items[i].VarName;
      StringGridVars.Cells[1,i+1] := '-INF';
      StringGridVars.Cells[2,i+1] := 'INF';
      StringGridVars.Cells[3,i+1] := iVariables.Items[i].Value.AsString;
      StringGridResults.Cells[0,i+1] := iVariables.Items[i].VarName;
    end;

  finally
    expr.Free;
    list.Free;
  end;
end;

procedure TfrmOptimBounded.SolveProblem;
var numvars: Integer;
  i: Integer;
  fMin: double;
  StopReason: TOptStopReason;
  expr: TMtxExpression;
  tol: double;
  maxiter: Integer;
begin
  expr := TMtxExpression.Create;
  Screen.Cursor := crHourGlass;
  try
    numvars := iVariables.Count;
    SetLength(pars,numvars);
    SetLength(LB,numvars);
    SetLength(UB,numvars);

    // #1: Setup expression evaluator
    for i := 0 to numvars-1 do
      expr.DefineDouble(iVariables[i].VarName).Value := StrToSample(StringGridVars.Cells[3,i+1]);

    expr.Expressions := EditFunction.Text;

    Memo1.Lines.Clear;
    // #2: Read values for optimization
    for i := 0 to numvars-1 do
    begin
      pars[i] := StrToSample(StringGridVars.Cells[3,i+1]);
      LB[i] := StrToSample(StringGridVars.Cells[1,i+1]);
      UB[i] := StrToSample(StringGridVars.Cells[2,i+1]);
    end;

    // #3: Setup and solve with Simplex
    tol := StrToSample(Edit1.Text);
    if tol < 1E-14 then tol := 1E-14;
    
    maxiter := UpDown1.Position;
    if CheckBox1.Checked then
      Simplex(OptObjFun,pars,[],[expr,iVariables],Lb,UB,fMin,StopReason,mvDouble, maxiter,tol,Memo1.Lines)
    else
      Simplex(OptObjFun,pars,[],[expr,iVariables],Lb,UB,fMin,StopReason,mvDouble, maxiter,tol);

    // #4: Write results
    Memo1.Lines.Add('');
    Case StopReason of
      OptResNotFound : Memo1.Lines.Add('Minimum not found.');
      OptResSmallGrad: Memo1.Lines.Add('Small gradient step.');
      OptResSingular:Memo1.Lines.Add('Singularity.');
      OptResSmallStep: Memo1.Lines.Add('Small step.');
      OptResZeroStep: Memo1.Lines.Add('Zero step.');
      OptResBigLambda: Memo1.Lines.Add('Big lambda.');
      OptResMaxIter: Memo1.Lines.Add('Maximum iteration limit reached.');
    else Memo1.Lines.Add('Converged within given tolerance.');
    End;

    Memo1.Lines.Add('Function value: '+SampleToStr(fMin,3,6));

    // Write results back to StringGridResults
    for i := 0 to numvars-1 do
      StringGridResults.Cells[1,i+1] := SampleToStr(pars[i],3,6);
  finally
    expr.Free;
    Screen.Cursor := crDefault;
  end;

end;

initialization
  RegisterClass(TfrmOptimBounded);

end.

File: MtxVec.CPP/Optim_Bounded.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#pragma hdrstop

#include "Optim_Bounded.h"
#include <MtxVec.hpp>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Basic2"
#pragma resource "*.dfm"
TfrmOptimBounded *frmOptimBounded;
//---------------------------------------------------------------------------
__fastcall TfrmOptimBounded::TfrmOptimBounded(TComponent* Owner)
  : TBasicForm2(Owner)
{
}
//---------------------------------------------------------------------------
static double __fastcall OptObjFun(TVec * const Parameters, TVec * const Constants,
                                   System::TObject* const *ObjConst,
                                   #if (__BORLANDC__ >= 0x0770)
                                   const System::NativeInt ObjConst_High
                                   #else
                                   const int ObjConst_Size
                                   #endif
                                   )
{
  TVariableCollection *vars = (TVariableCollection*)ObjConst[1];
  TMtxExpression *expr = (TMtxExpression*)ObjConst[0];
  // Typecast TVec* -> TVector* (MtxVec.h): pPars->Values[i] binds to TVector's inline
  // raw-pointer property (((double*)ValuesPointer)[i]), which shadows the inherited
  // DynamicArray Values FIELD. The field throws a false range error on the SetSubRange
  // views Simplex passes; the property is fast and correct. See CPP-toolchain.md.
  Mtxvec::TVector *pPars = (Mtxvec::TVector*) Parameters;
  for (int i = 0; i < pPars->Length; i++)
  {
    expr->VarByName->Value[vars->Items[i]->VarName]->DoubleValue = pPars->Values[i];
  }
  return expr->EvaluateDouble();
}
//---------------------------------------------------------------------------
void __fastcall TfrmOptimBounded::Button1Click(TObject *Sender) { SolveProblem(); }
void __fastcall TfrmOptimBounded::ButtonDetectClick(TObject *Sender) { InitSystem(); Button1->Enabled = true; }
void __fastcall TfrmOptimBounded::EditFunctionChange(TObject *Sender) { Button1->Enabled = false; }
//---------------------------------------------------------------------------
void __fastcall TfrmOptimBounded::FormCreate(TObject *Sender)
{
  EditFunction->Text = "100*Sqr(b-Sqr(a)) + Sqr(1-a)";
  iVariables = new TVariableCollection();

  RichEdit1->Lines->Clear();
  RichEdit1->Lines->Add(
    "Optimization now supports bounded constraints. Invididual parameter "
    "can have no, lower, upper or both limits. This is done by:");
  RichEdit1->Lines->Add("1. No limit: lower bound is -INF, upper bound is INF.");
  RichEdit1->Lines->Add("2. Lower bound: lower bound is lb, upper bound is INF.");
  RichEdit1->Lines->Add("3. Upper bound: lower bound is -INF, upper bound is ub.");
  RichEdit1->Lines->Add("4. Lower and upper bound: lower bound is lb, upper bound is ub.");
  RichEdit1->Lines->Add("");
  RichEdit1->Lines->Add("1. Enter any function in the \"Functon\" edit box.");
  RichEdit1->Lines->Add("2. Click the \"Detect\" button to get the list of variables in function.");
  RichEdit1->Lines->Add("3. Define bounds (if any) for individual variables.");
  RichEdit1->Lines->Add("4. Define starting values for individual variables.");
  RichEdit1->Lines->Add("5. Press the \"Calculate\" button to minimize function.");

  Edit1->Text = SampleToStr(1e-8);
  InitSystem();
  Button1->Enabled = true;
}
//---------------------------------------------------------------------------
void __fastcall TfrmOptimBounded::FormDestroy(TObject *Sender)
{
  delete iVariables;
  pars.Length = 0;
  LB.Length = 0;
  UB.Length = 0;
}
//---------------------------------------------------------------------------
void __fastcall TfrmOptimBounded::InitSystem()
{
  iVariables->Clear();
  TStringList *list = new TStringList();
  TMtxExpression *expr = new TMtxExpression();
  try {
    expr->AddExpr(EditFunction->Text);
    expr->GetVarList(list);

    StringGridVars->RowCount = list->Count + 1;
    StringGridResults->RowCount = list->Count + 1;
    StringGridVars->Cells[0][0] = "Name";
    StringGridVars->Cells[1][0] = "Lower bound";
    StringGridVars->Cells[2][0] = "Upper bound";
    StringGridVars->Cells[3][0] = "Initial value";
    StringGridResults->Cells[0][0] = "Name";
    StringGridResults->Cells[1][0] = "Value";

    for (int i = 0; i < list->Count; i++)
    {
      TNamedVariable *v = iVariables->Add(list->Strings[i]);
      v->Value->DefineDouble();
    }

    for (int i = 0; i < iVariables->Count; i++)
    {
      StringGridVars->Cells[0][i + 1] = iVariables->Items[i]->VarName;
      StringGridVars->Cells[1][i + 1] = "-INF";
      StringGridVars->Cells[2][i + 1] = "INF";
      StringGridVars->Cells[3][i + 1] = iVariables->Items[i]->Value->AsString;
      StringGridResults->Cells[0][i + 1] = iVariables->Items[i]->VarName;
    }
  } __finally {
    delete expr;
    delete list;
  }
}
//---------------------------------------------------------------------------
void __fastcall TfrmOptimBounded::SolveProblem()
{
  TMtxExpression *expr = new TMtxExpression();
  Screen->Cursor = crHourGlass;
  try {
    int numvars = iVariables->Count;
    pars.Length = numvars;
    LB.Length = numvars;
    UB.Length = numvars;

    for (int i = 0; i < numvars; i++)
      expr->DefineDouble(iVariables->Items[i]->VarName)->Value = StrToSample(StringGridVars->Cells[3][i + 1]);
    expr->Expressions = EditFunction->Text;

    Memo1->Lines->Clear();
    for (int i = 0; i < numvars; i++)
    {
      pars[i] = StrToSample(StringGridVars->Cells[3][i + 1]);
      LB[i] = StrToSample(StringGridVars->Cells[1][i + 1]);
      UB[i] = StrToSample(StringGridVars->Cells[2][i + 1]);
    }

    double tol = StrToSample(Edit1->Text);
    if (tol < 1E-14) tol = 1E-14;
    int maxiter = UpDown1->Position;
    double fMin;
    TOptStopReason StopReason;

    if (CheckBox1->Checked)
      Simplex(OptObjFun, &pars[0], pars.Length - 1,
              NULL, -1,
              OPENARRAY(TObject*, (expr, iVariables)),
              &LB[0], LB.Length - 1, &UB[0], UB.Length - 1,
              fMin, StopReason, mvDouble, maxiter, tol, Memo1->Lines);
    else
      Simplex(OptObjFun, &pars[0], pars.Length - 1,
              NULL, -1,
              OPENARRAY(TObject*, (expr, iVariables)),
              &LB[0], LB.Length - 1, &UB[0], UB.Length - 1,
              fMin, StopReason, mvDouble, maxiter, tol);

    Memo1->Lines->Add("");
    switch (StopReason)
    {
      case OptResNotFound:  Memo1->Lines->Add("Minimum not found."); break;
      case OptResSmallGrad: Memo1->Lines->Add("Small gradient step."); break;
      case OptResSingular:  Memo1->Lines->Add("Singularity."); break;
      case OptResSmallStep: Memo1->Lines->Add("Small step."); break;
      case OptResZeroStep:  Memo1->Lines->Add("Zero step."); break;
      case OptResBigLambda: Memo1->Lines->Add("Big lambda."); break;
      case OptResMaxIter:   Memo1->Lines->Add("Maximum iteration limit reached."); break;
      default:              Memo1->Lines->Add("Converged within given tolerance."); break;
    }
    Memo1->Lines->Add("Function value: " + SampleToStr(fMin, 3, 6));

    for (int i = 0; i < numvars; i++)
      StringGridResults->Cells[1][i + 1] = SampleToStr(pars[i], 3, 6);
  } __finally {
    delete expr;
    Screen->Cursor = crDefault;
  }
}
//---------------------------------------------------------------------------
